public/app/plugins/datasource/prometheus/querybuilder/shared/parsingUtils.ts

```diff
@@ -0,0 +1,179 @@
+  return {
+    text: getString(expr, node),
+    // TODO: this are positions in the string with the replaced variables. Means it cannot be used to show exact
+    //  placement of the error for the user. We need some translation table to positions before the variable
+    //  replace.
+    from: node.from,
+    to: node.to,
+    parentType: node.parent?.name,
+  };
+}
+
+// Taken from template_srv, but copied so to not mess with the regex.index which is manipulated in the service
+/*
+ * This regex matches 3 types of variable reference with an optional format specifier
+ * \$(\w+)                          $var1
+ * \[\[([\s\S]+?)(?::(\w+))?\]\]    [[var2]] or [[var2:fmt2]]
+ * \${(\w+)(?::(\w+))?}             ${var3} or ${var3:fmt3}
+ */
+const variableRegex = /\$(\w+)|\[\[([\s\S]+?)(?::(\w+))?\]\]|\${(\w+)(?:\.([^:^\}]+))?(?::([^\}]+))?}/g;
+
+/**
+ * As variables with $ are creating parsing errors, we first replace them with magic string that is parseable and at
+ * the same time we can get the variable and it's format back from it.
+ * @param expr
+ */
+export function replaceVariables(expr: string) {
+  return expr.replace(variableRegex, (match, var1, var2, fmt2, var3, fieldPath, fmt3) => {
+    const fmt = fmt2 || fmt3;
+    let variable = var1;
+    let varType = '0';
+
+    if (var2) {
+      variable = var2;
+      varType = '1';
+    }
+
+    if (var3) {
+      variable = var3;
+      varType = '2';
+    }
+
+    return `__V_${varType}__` + variable + '__V__' + (fmt ? '__F__' + fmt + '__F__' : '');
+  });
+}
+
+const varTypeFunc = [
+  (v: string, f?: string) => `\$${v}`,
+  (v: string, f?: string) => `[[${v}${f ? `:${f}` : ''}]]`,
+  (v: string, f?: string) => `\$\{${v}${f ? `:${f}` : ''}\}`,
+];
+
+/**
+ * Get beck the text with variables in their original format.
+ * @param expr
+ */
+function returnVariables(expr: string) {
+  return expr.replace(/__V_(\d)__(.+)__V__(?:__F__(\w+)__F__)?/g, (match, type, v, f) => {
+    return varTypeFunc[parseInt(type, 10)](v, f);
+  });
+}
+
+/**
+ * Get the actual string of the expression. That is not stored in the tree so we have to get the indexes from the node
+ * and then based on that get it from the expression.
+ * @param expr
+ * @param node
+ */
+export function getString(expr: string, node: SyntaxNode | TreeCursor | null | undefined) {
+  if (!node) {
+    return '';
+  }
+  return returnVariables(expr.substring(node.from, node.to));
+}
+
+/**
+ * Create simple scalar binary op object.
+ * @param opDef - definition of the op to be created
+ * @param expr
+ * @param numberNode - the node for the scalar
+ * @param hasBool - whether operation has a bool modifier. Is used only for ops for which it makes sense.
+ */
+export function makeBinOp(
+  opDef: { id: string; comparison?: boolean },
+  expr: string,
+  numberNode: SyntaxNode,
+  hasBool: boolean
+) {
+  const params: any[] = [parseFloat(getString(expr, numberNode))];
+  if (opDef.comparison) {
+    params.unshift(hasBool);
```

Comment:
is there a test case for this one? i'm not following how it's used in a query

Reply:
You mean this particular param? There are few tests that test the general parsing so I didn't think this needs a separate test. For example here https://github.com/grafana/grafana/blob/20b4112b3b4ebba355c87aea1a53e3c098a36bd3/public/app/plugins/datasource/prometheus/querybuilder/parsing.test.ts#L432

The bool param is a specific thing that basically says what is the result of binary operators like `/`. https://prometheus.io/docs/prometheus/latest/querying/operators/#comparison-binary-operators